### PR TITLE
MULE-16527: Remove `flatMap` in UntilSuccesful

### PR DESCRIPTION
**Pull request by plbalbi**

### CODE REVIEW

**Comment by plbalbi** [comment]
--validate

**Comment by plbalbi** [comment]
--validate

**Comment by plbalbi** [comment]
--validare

**Comment by plbalbi** [comment]
--validate

**Comment by plbalbi** [comment]
--validate

File: core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    // Downstream chain. Unpacks and publishes successful events and errors downstream.
+    downstreamFlux = Flux.create(downstreamRecorder)
+        .doOnNext(event -> inflightEvents.decrementAndGet())
+        .map(getScopeResultMapper());
+
+    if (expressionManager.isExpression(maxRetries)) {
+      maxRetriesSupplier = expressionToIntegerSupplierFor(maxRetries);
+    } else {
+      Integer asInteger = Integer.parseInt(maxRetries);
+      maxRetriesSupplier = session -> asInteger;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java]: inline + static import

File: core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    }
+
+    if (expressionManager.isExpression(millisBetweenRetries)) {
+      delaySupplier = expressionToIntegerSupplierFor(millisBetweenRetries);
+    } else {
+      Integer asInteger = Integer.parseInt(millisBetweenRetries);
+      delaySupplier = session -> asInteger;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java]: inline + static import

File: core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+    RetryContext(CoreEvent event) {
+      this.event = event;
+
+      ExpressionManagerSession session = sessionSupplier.apply(event);
+      maxRetries = maxRetriesSupplier.apply(session);
+      delayInMillis = delaySupplier.apply(session);
+      retryCount.set(maxRetries);
+
+      delayScheduler =
+          new ConditionalExecutorServiceDecorator(UntilSuccessfulRouter.this.delayScheduler, s -> isTransactionActive());
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java]: the delay scheduler from `UntilSuccessfulRouter ` is already decorated

File: core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+   * Context carrying all retrials information.
+   */
+  class RetryContext {
+
+    CoreEvent event;
+    AtomicInteger retryCount = new AtomicInteger();
+
+    Integer delayInMillis;
+    Integer maxRetries;
+    ConditionalExecutorServiceDecorator delayScheduler;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java]: why do you need this field here?

File: core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+   * 
+   * @param event the current retrial {@link CoreEvent}
+   * @param ctx the current {@link RetryContext}
+   * @return the {@link CoreEvent} with the retry context saved as internal parameter
+   */
+  private CoreEvent pushRetryContext(CoreEvent event, RetryContext ctx) {
+    // Requires: The ctx that corresponds to this router execution is not in the stack, or there's no stack yet
+    // Assures: The ctx that corresponds to this router execution is the one on top of the stack
+
+    // TODO: There must be cleaner way to do this
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java]: i has a similar issue for handling nested policies, we should extract that and reuse here

File: core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+    // Check if I'm already inside another until-successful scope
+    currentStack = ((InternalEvent) event).getInternalParameter(RETRY_CTX_INTERNAL_PARAM_KEY);
+    if (currentStack == null) {
+      // First scope
+      currentStack = new Stack<>();
+    }
+
+    currentStack.push(ctx);
+
+    Map<String, Object> parametersWithCtx = new HashMap<>();
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java]: use `Collections.singletonMap`

File: core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+      RetryContext ctx = getRetryContextForEvent(messagingError.getEvent());
+      int retriesLeft =
+          ctx.retryCount.getAndDecrement();
+      if (retriesLeft > 0) {
+        LOGGER.error("Retrying execution of event, attempt {} of {}.", ctx.getRetriesLeft(),
+                     ctx.maxRetries != RETRY_COUNT_FOREVER ? ctx.maxRetries : "unlimited");
+
+        // Schedule retry with delay
+        UntilSuccessfulRouter.this.delayScheduler.schedule(() -> innerRecorder.next(eventWithCurrentContext(ctx.event, ctx)),
+                                                           ctx.delayInMillis, TimeUnit.MILLISECONDS);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java]: static import

File: core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+  }
+
+  /**
+   * Assembles and returns the downstream {@link Publisher<CoreEvent>}.
+   * 
+   * @return the successful {@link CoreEvent} or retries exhaustion errors {@link Publisher}
+   */
+  Publisher<CoreEvent> getDownstreamPublisher() {
+    return downstreamFlux
+        .compose(downstreamPublisher -> Mono.subscriberContext()
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java]: static import

**Pull request merged**